### PR TITLE
Added OLED fade out support

### PR DESCRIPTION
**Pull request by Barabas5532**
## Description
Instead of turning off instantly, the screen gradually fades to black.

See issue #11433.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by Barabas5532** [comment]
I'm not sure if there are any tests for the oled driver. Running `make test:all` fails and `test:oled` `test:driver` doesn't run any tests.

**Review by drashna — APPROVED** [review]
Tested it out and looks good. 

Only issue is that it suddenly turns off before fading all the way out. But I suspect that's a problem with the OLED itself, and not what is done here (since there is almost nothing done here)

**Comment by Barabas5532** [comment]
> Only issue is that it suddenly turns off before fading all the way out.

I've noticed this too. It is due to the OLED still being quite bright at the dimmest setting. The last step before it turns off is equivalent to setting the brightness to 0. Not much that can be done about this unfortunately.

File: drivers/oled/oled_driver.c [hunk 1]
```diff
@@ -71,6 +71,9 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #define PRE_CHARGE_PERIOD 0xD9
 #define VCOM_DETECT 0xDB
 
+// Advance Graphic Commands
+#define FADE_BLINK 0x23
```
Comment (XScorpion2) [inline drivers/oled/oled_driver.c:75]: I can't find any reference to this command in the spec sheet. Able to link me to where you found this info?

Comment (XScorpion2) [reply drivers/oled/oled_driver.c:75]: Whelp, just found it right after posting it.

File: drivers/oled/oled_driver.c [hunk 2]
```diff
@@ -555,7 +564,13 @@ bool oled_off(void) {
         return !oled_active;
     }
 
-    static const uint8_t PROGMEM display_off[] = {I2C_CMD, DISPLAY_OFF};
+    static const uint8_t PROGMEM display_off[] =
+#ifdef OLED_DRIVER_DISPLAY_FADE
+        {I2C_CMD, FADE_BLINK, 0x20 | OLED_DRIVER_DISPLAY_FADE_INTERVAL};
```
Comment (XScorpion2) [inline drivers/oled/oled_driver.c:569]: Please add a constant for '0x20' so we know this is the "Enable Fade Out mode." flag.
Additionally, might be useful to add beside that define, a '0x30' define for blinking.

**Review by XScorpion2 — CHANGES_REQUESTED** [review]
Super cool find! Small suggestions for code readability.

File: drivers/oled/oled_driver.c (hunk 1 above)
Comment (drashna) [reply drivers/oled/oled_driver.c:75]: > Whelp, just found it right after posting it.

Naturally. :D

File: docs/feature_oled_driver.md [hunk 3]
```diff
@@ -136,21 +136,23 @@ void oled_task_user(void) {
...
+|`OLED_FONT_END`                    |`223`            |The ending character index for custom fonts                                                                               |
+|`OLED_FONT_WIDTH`                  |`6`              |The font width                                                                                                            |
+|`OLED_FONT_HEIGHT`                 |`8`              |The font height (untested)                                                                                                |
+|`OLED_TIMEOUT`                     |`60000`          |Turns off the OLED screen after 60000ms of keyboard inactivity. Helps reduce OLED Burn-in. Set to 0 to disable.           |
+|`OLED_DRIVER_DISPLAY_FADE`         |*Not defined*    |Enables fade out animation. Use together with `OLED_TIMEOUT`.                                                             |
+|`OLED_DRIVER_DISPLAY_FADE_INTERVAL`|`0`              |The speed of fade out animation, from 0 to 15. Larger values are slower.                                                  |
```
Comment (fauxpark) [inline docs/feature_oled_driver.md:149]:
```suggestion
|`OLED_FADE_OUT`         |*Not defined*    |Enables fade out animation. Use together with `OLED_TIMEOUT`.                                                             |
|`OLED_FADE_OUT_INTERVAL`|`0`              |The speed of fade out animation, from 0 to 15. Larger values are slower.                                                  |
```

File: drivers/oled/oled_driver.c [hunk 4]
```diff
@@ -539,7 +542,13 @@ bool oled_on(void) {
     oled_timeout = timer_read32() + OLED_TIMEOUT;
 #endif
 
-    static const uint8_t PROGMEM display_on[] = {I2C_CMD, DISPLAY_ON};
+    static const uint8_t PROGMEM display_on[] =
+#ifdef OLED_DRIVER_DISPLAY_FADE
```
Comment (fauxpark) [inline drivers/oled/oled_driver.c:546]:
```suggestion
#ifdef OLED_FADE_OUT
```

File: drivers/oled/oled_driver.c [hunk 5]
```diff
@@ -555,7 +564,13 @@ bool oled_off(void) {
         return !oled_active;
     }
 
-    static const uint8_t PROGMEM display_off[] = {I2C_CMD, DISPLAY_OFF};
+    static const uint8_t PROGMEM display_off[] =
+#ifdef OLED_DRIVER_DISPLAY_FADE
```
Comment (fauxpark) [inline drivers/oled/oled_driver.c:568]:
```suggestion
#ifdef OLED_FADE_OUT
```

File: drivers/oled/oled_driver.c (hunk 2 above)
Comment (fauxpark) [inline drivers/oled/oled_driver.c:569]:
```suggestion
        {I2C_CMD, FADE_BLINK, 0x20 | OLED_FADE_OUT_INTERVAL};
```

File: drivers/oled/oled_driver.h [hunk 6]
```diff
@@ -154,6 +154,14 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #    endif
 #endif
 
+#if !defined(OLED_DRIVER_DISPLAY_FADE_INTERVAL)
+#    define OLED_DRIVER_DISPLAY_FADE_INTERVAL 0x00
+#endif
+
+#if OLED_DRIVER_DISPLAY_FADE_INTERVAL > 0x0F || OLED_DRIVER_DISPLAY_FADE_INTERVAL < 0x00
+#    error OLED_DRIVER_DISPLAY_FADE_INTERVAL must be between 0x00 and 0x0F
```
Comment (fauxpark) [inline drivers/oled/oled_driver.h:162]:
```suggestion
#if !defined(OLED_FADE_OUT_INTERVAL)
#    define OLED_FADE_OUT_INTERVAL 0x00
#endif

#if OLED_FADE_OUT_INTERVAL > 0x0F || OLED_FADE_OUT_INTERVAL < 0x00
#    error OLED_FADE_OUT_INTERVAL must be between 0x00 and 0x0F
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: docs/feature_oled_driver.md (hunk 3 above)
Comment (Barabas5532) [reply docs/feature_oled_driver.md:149]: @drashna, as you proposed these names, could you comment on this suggestion?

The new names would be more consistent with the existing options for the OLED driver. Here is an excerpt from the docs (using the exisiting names):
|Define                             |Default          |Description                                                                                                               |
|-----------------------------------|-----------------|--------------------------------------------------------------------------------------------------------------------------|                                                           
|`OLED_FONT_WIDTH`                  |`6`              |The font width                                                                                                            |
|`OLED_FONT_HEIGHT`                 |`8`              |The font height (untested)                                                                                                |
|`OLED_TIMEOUT`                     |`60000`          |Turns off the OLED screen after 60000ms of keyboard inactivity. Helps reduce OLED Burn-in. Set to 0 to disable.           |
|`OLED_DRIVER_DISPLAY_FADE`         |*Not defined*    |Enables fade out animation. Use together with `OLED_TIMEOUT`.                                                             |
|`OLED_DRIVER_DISPLAY_FADE_INTERVAL`|`0`              |The speed of fade out animation, from 0 to 15. Larger values are slower.

File: drivers/oled/oled_driver.c (hunk 2 above)
Comment (Barabas5532) [reply drivers/oled/oled_driver.c:569]: Do you have any preferences for naming? I was thinking something like this, but it's not that obvious that the `ENABLE_FADE` and `ENABLE_BLINK` are supposed to follow the `FADE_BLINK` command: 

```
// Advance Graphic Commands
#define FADE_BLINK 0x23
#define ENABLE_FADE 0x20
#define ENABLE_BLINK 0x30
```

```
{I2C_CMD, FADE_BLINK, ENABLE_FADE | OLED_DRIVER_DISPLAY_FADE_INTERVAL};
```

Comment (XScorpion2) [reply drivers/oled/oled_driver.c:569]: Ya, that looks fine to me.

File: docs/feature_oled_driver.md (hunk 3 above)
Comment (XScorpion2) [reply docs/feature_oled_driver.md:149]: Ya, drop the  _DRIVER portion of those names as Faux mentioned. When I wrote all the original defines, I dropped the _DRIVER as it was verbose, annoying to continually type, and didn't add a whole lot of clarity to the define.

**Review by fauxpark — APPROVED** [review]

**Comment by Barabas5532** [comment]
Looks like Travis CI failed due to unrelated issues with the develop branch. That seems sorted now, is there a way to restart the job?

**Comment by fauxpark** [comment]
You may need to rebase.

**Review by Erovia — APPROVED** [review]
`__attribute__ ((weak))`

**Pull request merged**